ls1intum/Artemis
# `Programming Exercises`: Experimental support for GitLab CI

### Checklist
#### General
- [x] I tested **all** changes and their related features on the GitLab CI test server.
#### Server
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features on the GitLab CI test server.

### Motivation and Context
We have addressed the following issues which arrise when using the current implementation with either Bamboo/Bitbucket and Jenkins/GitLab:
- When conducting programming exercises, Artemis relies on a Version Control System (VCS) as well as a Continuous Integration System (CIS). Right now, it is required to setup and maintain two systems. As some VCSs also provide an in-built Continuous Integration (CI) applicaton (e.g. GitHub with GitHub Actions and GitLab with GitLab CI), it is possible to simplify the system landscape and to depend only on one system, handling the Version Control and CI aspect.
- For generating the feedback by executing the tests the CIS needs a configuration. This configuration is duplicated for each participation in a programming exercise. First of all, over time this leads to a performance loss in Bamboo and Jenkins. Secondly, it is hard to update the build plan throughout the exercise lifecycle. By introducing the build plan API in Artemis, the build plans are centrally stored once and fetched by the CIS on each run. This allows instructors to update the build plan easily.

After evaluating different CISs, we came to the conclustion, that in this context GitLab is the best option.

Furthermore, we have generalized the existing notification plugin for Jenkins by adapting the system design.
We have published the new, system independent plugin on GitHub: https://github.com/ls1intum/artemis-notification-plugin

We have developed an experimental version of Artemis, which is now capable of conducting Java programming exercises with a single dependency on GitLab by using the integrated CI.

### Description
This pull requests consists mainly of the following pull requests:
- https://github.com/ls1intum/Artemis/pull/5021
- https://github.com/ls1intum/Artemis/pull/5658
- https://github.com/ls1intum/Artemis/pull/5807
- https://github.com/ls1intum/Artemis/pull/5808

### Future ToDos
The following aspects are part of future work:
- unique index on the hash of the build plans stored in the database
- updating the CI variables
- setting CI variables on the group level
- usage of a GitLab access token which is only valid for the test repository
- encrypting the build artifacts in GitLab
- setting a build timeout
- exams
- static code analysis
- testwise coverage report
- import of programming exercises
- viewing build logs in the online editor
- trigger of the build on creation of a new exercise
- sending build results of solution and test repository
- usage of the custom feedback directory
- improve overall performance regarding the API calls
- link to build plans in exercise overview
- remove log error messages
- calculate build duration

### Steps for Testing
`Note`: You cannot test this PR on one of the test servers. You either have to use the local setup or send me a message via Slack or Email and I can provide you access to a special GitLab CI test server.

`Local Setup Instructions`: https://artemis-platform--6044.org.readthedocs.build/en/6044/dev/setup/#gitlab-ci-and-gitlab-setup

Please test the "normal" programming exercise workflow. Additional features (static code analysis etc.) and exams are not implemented yet (see Todos).

1. Create a new programming exercise
2. Check if the template, solution and test repository are created in GitLab
3. Start the programming exercise
4. Open the repository in GitLab
5. Access the build plan with the URL you find in  `Settings` > `CI/CD` > `General Pipelines` > `CI/CD configuration file`
6. Commit changes to the code
7. Check if the pipeline is running in GitLab
8. View the feedback in Artemis

### Review Progress

#### Code Review
#### Manual Tests

### Test Coverage
